src/main/java/bio/terra/cli/Resource.java

```diff
@@ -0,0 +1,229 @@
+import bio.terra.cli.resources.GcsBucket;
+import bio.terra.cli.serialization.command.CommandResource;
+import bio.terra.cli.serialization.disk.DiskResource;
+import bio.terra.cli.service.WorkspaceManagerService;
+import bio.terra.workspace.model.AccessScope;
+import bio.terra.workspace.model.CloningInstructionsEnum;
+import bio.terra.workspace.model.ControlledResourceIamRole;
+import bio.terra.workspace.model.ControlledResourceMetadata;
+import bio.terra.workspace.model.ManagedBy;
+import bio.terra.workspace.model.PrivateResourceUser;
+import bio.terra.workspace.model.ResourceDescription;
+import bio.terra.workspace.model.ResourceMetadata;
+import bio.terra.workspace.model.StewardshipType;
+import java.util.List;
+import java.util.UUID;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+
+/**
+ * Internal representation of a workspace resource. This abstract class contains properties common
+ * to all resource types. Sub-classes represent a specific resource type. Instances of the
+ * sub-classes are part of the current context or state.
+ */
+public abstract class Resource {
+  // all resources
+  protected UUID id;
+  protected String name;
+  protected String description;
+  protected Type resourceType;
+  protected StewardshipType stewardshipType;
+  protected CloningInstructionsEnum cloningInstructions;
+
+  // controlled resources
+  protected AccessScope accessScope;
+  protected ManagedBy managedBy;
+
+  // private controlled resources
+  protected String privateUserName;
+  protected List<ControlledResourceIamRole> privateUserRoles;
+
+  /**
+   * Enum for the types of workspace resources supported by the CLI. Each enum value maps to a
+   * single WSM client library ({@link bio.terra.workspace.model.ResourceType}) enum value.
+   *
+   * <p>The CLI defines its own enum instead of using the WSM one so that we can restrict the
+   * resource types supported (e.g. no Data Repo snapshots). It also gives the CLI control over what
+   * the enum names are, which are exposed to users as command options.
+   */
+  public enum Type {
+    GCS_BUCKET,
+    BQ_DATASET,
+    AI_NOTEBOOK;
+  }
+
+  /** Deserialize an instance of the disk format to the internal object. */
+  protected Resource(DiskResource configFromDisk) {
+    this.id = configFromDisk.id;
+    this.name = configFromDisk.name;
+    this.description = configFromDisk.description;
+    this.resourceType = configFromDisk.resourceType;
+    this.stewardshipType = configFromDisk.stewardshipType;
+    this.cloningInstructions = configFromDisk.cloningInstructions;
+    this.accessScope = configFromDisk.accessScope;
+    this.managedBy = configFromDisk.managedBy;
+    this.privateUserName = configFromDisk.privateUserName;
+    this.privateUserRoles = configFromDisk.privateUserRoles;
+  }
+
+  /** Deserialize an instance of the WSM client library object to the internal object. */
+  protected Resource(ResourceMetadata metadata) {
```

Comment:
What do you mean "created by other means"? We create an instance of this class either by making a call to WSM to create a new resource or by reading in a cached version from disk. Is there another way I'm missing?

If I use a builder here, I think every new conversion would just require a new method on the builder, instead of a new constructor. So not sure what's saved by taking one approach over the other.

RE adding more pojos: Not sure I follow your suggestion here. Are you thinking of an intermediary pojo (e.g. disk pojo -> creation arguments pojo -> internal obj), with the goal of sharing a single constructor for the disk/wsm pojo deserializations? If so, I'm wary of adding another de/serialization step just to save a constructor.

Reply:
clarifying, but not still optional:
"created by other means": Making every means a specific constructor means we have to modify this class everytime how we create instances changes. These constructors know too much about how a Resource is created, but we should be more agnostic. Also source of the circular dependencies. We're not creating it by any other means now, but a constructor per means makes future changes more invasive than they need to be.

If we use a builder, I would similarly say the builder should not have methods on it that are aware of the means of construction e.g. `ResourcMetadata`. Instead, have fields for each.

Yes, the code to convert from a `ResourceMetadata` -> the arguments to create the Resource still exists somewhere, but it's better for it to exist on or adjacent to `ResourceMetadata` instead of putting so much on `Resource`.

Re more pojos: An intermediate *value* class, no serialization involved. Because Java doesn't have records (until JDK 14), that's a pojo or an `@AutoValue`, but we don't need or want to serialize in the middle.
Normally a builder would be sufficient, but with inheritance builders get more exotic as ou need subclass builders that are aware of the type they're creating. Either a builder of creation arguments value class are ways to make the construction of `Resource` more readable than 10 argument constructor lists, but that's an added refinement that's not necessary.

Reply:
Gotcha, thanks for the clarification. I'm going to leave this as is for now, and we can tackle it when we introduce another serialization format and/or way to create an instance of these objects.

RE more pojos: I guess a builder is really just another intermediate value class as well, and I wouldn't necessarily be against that.